Enable Caffe2's Sum Support Broadcasting

zqhsfz2 opened the pull request:
Summary:
Enable Caffe2's Sum Support Broadcasting. This involves several following changes:

1. Modify SumOp implementation such that it supports broadcasting
2. Add SumGradientOp implementation and register with SumOp

Differential Revision: D13834042

zqhsfz2 commented on caffe2/operators/utility_ops.cu:
```diff
@@ -36,12 +36,27 @@ bool SumOp<CUDAContext>::RunOnDevice() {
 
+template <>
+bool SumGradientOp<CUDAContext>::RunOnDevice() {
+  CAFFE_ENFORCE_GT(InputSize(), 1);
```
Since this is inside SumGradientOp instead of SumOp, the first input corresponds to dY (Y being the output of SumOp) and second input corresponds to first input of SumOp, right?

houseroad replied:
@zqhsfz2 could you explain in your SumGradient, what are input(s) and output(s)?

zqhsfz2 replied:
Sure. Assuming what SumOp does is Y = X1 + X2 + X3 + ... Xn, the inputs for SumGradientOp would be (dY, X1, X2, ..., Xn) and outputs for SumGradientOp would be (dX1, dX2, ..., dXn). This has been specified in L19-25 of:

https://github.com/pytorch/pytorch/pull/16485/files#diff-1e7bf4989a4beddf9c0c537e29f47ef0R16

houseroad commented on caffe2/operators/elementwise_sum_op.cc:
```diff
@@ -124,4 +143,11 @@ A after Sum: [[10.  7. 11.]
     .InheritOnnxSchema();
-}
+
+OPERATOR_SCHEMA(SumGradient)
```
Could you also update the doc of the Sum operator to show that it supports broadcasting?

houseroad commented on caffe2/operators/utility_ops.cu:
```diff
@@ -36,12 +36,27 @@ bool SumOp<CUDAContext>::RunOnDevice() {
+  } else if (Input(1).IsType<at::Half>()) {
+    return DoRunWithType<at::Half>();
+  } else {
+    CAFFE_THROW("Unsupported inputs");
```
Could you also print out the unsupported data type?

houseroad commented on caffe2/operators/utility_ops.h:
```diff
@@ -266,33 +266,60 @@ class SumOp : public Operator<Context> {
+    // initialize output
     auto* output = Output(0, input0.sizes(), at::dtype<T>());
     T* output_data = output->template mutable_data<T>();
+    std::vector<int> output_dims = GetDimsVec(*output);
```
Don't we need to compute the new output size based on the broadcasting? And resize the output?

houseroad commented on caffe2/operators/utility_ops.h:
```diff
@@ -266,33 +266,60 @@ class SumOp : public Operator<Context> {
+      }
+
+      // dim vec of input i should be suffix of output dim vec
+      if (!std::equal(
```
Why do need this check? [2, 1, 2] is not suffix of [2, 2, 2, 2]?

houseroad reviewed (changes requested):
I think the understanding of the numpy-broadcasting is incorrect, and we should resize the output first.

houseroad commented on caffe2/operators/elementwise_sum_op.cc:
```diff
@@ -26,7 +45,15 @@ OPERATOR_SCHEMA(Sum)
+
+This operation supports numpy-style broadcasting when shape between inputs is
+different. Essentially the shape between two tensors would be compared starting
+from training dimensions, moving itsway forward. Two dimensions are compatible
```
training dimensions?

houseroad commented on caffe2/operators/utility_ops.cu:
```diff
@@ -31,7 +31,21 @@ bool SumOp<CUDAContext>::RunOnDevice() {
+
+  if (Input(1).IsType<float>()) {
+    return DoRunWithType<float>();
+  } else if (Input(1).IsType<at::Half>()) {
```
Why do we want to add Half in this case?

houseroad commented on caffe2/operators/utility_ops.h:
```diff
@@ -256,48 +256,90 @@ class SumOp : public Operator<Context> {
+
+    // check for in-place case
+    if (IsInputOutputAlias(0, 0)) {
+      CAFFE_ENFORCE(
```
Why not just use CAFFE_EFORCE_EQ like what you did before?

houseroad commented on caffe2/operators/utility_ops.h:
```diff
@@ -313,6 +355,98 @@ class SumOp : public Operator<Context> {
   }
+
+ private:
+  std::vector<int64_t> GetBroadcastDims(
```
Can we move this function to elementwise_ops_utils?

houseroad commented on caffe2/operators/utility_ops.h:
```diff
@@ -313,6 +355,98 @@ class SumOp : public Operator<Context> {
+        int64_t dim_value_cur = *(input_dims.rbegin() + i);
+
+        CAFFE_ENFORCE(
+            dim_value == dim_value_cur || dim_value <= 1 || dim_value_cur <= 1);
```
Provide some useful error message here? It will help us debug the problem.

houseroad reviewed (commented):
Do you want to rebase to master and collect some signals from CI?

zqhsfz2 commented:
> Do you want to rebase to master and collect some signals from CI?

Yeah, I find that there is quite a lot of changes in the master so I would rebase and adapt accordingly.

houseroad commented on caffe2/operators/utility_ops.cc:
```diff
@@ -726,17 +726,6 @@ REGISTER_GRADIENT(Alias, GetAliasGradient);
 
 SHOULD_NOT_DO_GRADIENT(ResizeLike);
 
-class GetSumGradient : public GradientMakerBase {
```
hmm, we also have an old SumGradient here... We need to make the new one backward compatible with the old one. To achieve this, you probably need to check the input number and output number in the gradient op, and decide which version of the operator we are handling...

houseroad commented on caffe2/operators/utility_ops.cu:
```diff
@@ -31,7 +31,19 @@ bool SumOp<CUDAContext>::RunOnDevice() {
+
+  if (Input(1).IsType<float>()) {
+    return DoRunWithType<float>();
+  } else {
```
I think we need to support half in the gradient as well.

zqhsfz2 replied:
OK ...

houseroad commented on caffe2/operators/utility_ops.h:
```diff
@@ -315,6 +356,59 @@ class SumOp : public Operator<Context> {
+
+    if (Input(1).template IsType<float>()) {
+      return DoRunWithType<float>();
+    } else if (Input(1).template IsType<int>()) {
```
The original version support int or not?

zqhsfz2 replied:
In original version, SumOp support float and int in CPU. So perhaps we stick with that?

houseroad reviewed (commented):
let's make it clear what types should be supported.
GPU: float and half
CPU: float, (int?)

BIT-silence commented on caffe2/operators/elementwise_ops_utils.cc:
```diff
@@ -71,6 +71,50 @@ std::vector<int> ComputeBinaryBroadcastForwardDims(
+    const std::vector<std::vector<int>>& input_dims_list) {
+  // get output rank first
+  int output_rank = 0;
+  for (auto input_dims : input_dims_list) {
```
```suggestion
  for (const auto& input_dims : input_dims_list) {
```

BIT-silence commented on caffe2/operators/elementwise_ops_utils.cc:
```diff
@@ -71,6 +71,50 @@ std::vector<int> ComputeBinaryBroadcastForwardDims(
+  std::vector<int> output_dims;
+  for (int i = 0; i < output_rank; i++) {
+    int dim_value = 1;
+    for (auto input_dims : input_dims_list) {
```
```suggestion
    for (const auto& input_dims : input_dims_list) {
```

BIT-silence commented on caffe2/utils/math/reduce.cu:
```diff
@@ -533,6 +517,7 @@ DELEGATE_CUDA_REDUCE_FUNCTION(std::int32_t, ReduceSum, cub::Sum, 0)
 DELEGATE_CUDA_REDUCE_FUNCTION(std::int64_t, ReduceSum, cub::Sum, 0LL)
 DELEGATE_CUDA_REDUCE_FUNCTION(float, ReduceSum, cub::Sum, 0.0f)
 DELEGATE_CUDA_REDUCE_FUNCTION(double, ReduceSum, cub::Sum, 0.0)
+DELEGATE_CUDA_REDUCE_FUNCTION(at::Half, ReduceSum, cub::Sum, at::Half(0.0))
```
For ReduceSum part, I'd suggest write another specialization function. Since by default for half type computation, we will convert it to float32 and then convert back, while we cannot assume that cub::Sum will do this.

BIT-silence commented on caffe2/utils/math_gpu.cu:
```diff
@@ -2911,11 +2911,7 @@ __global__ void BroadcastCUDAKernel(
       FIXED_DIVISOR_DIV_MOD(Y_dims.data[i], Y_index_val, &Y_index_val, &d);
       X_index += d * X_strides.data[i];
     }
```
ditto

Commit:
Enable Caffe2's Sum Support Broadcasting (#16485)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/16485

Enable Caffe2's Sum Support Broadcasting. This involves several following changes:

1. Modify SumOp implementation such that it supports broadcasting
2. Add SumGradientOp implementation and register with SumOp

Differential Revision: D13834042

fbshipit-source-id: 63146bfdae05ac6d0521b354256accb6a094f832
